rfcs/error-identifiers.md

```diff
@@ -0,0 +1,176 @@
+# Short identifiers for error messages
+
+## Motivation
+
+We propose to assign stable, short identifiers to each error produced
+by the OCaml compiler, and include this short name in the error
+message. This makes it much easier to refer to those errors message in
+documentation, user questions / answers, and to search on the web for
+material on a specific error.
+
+This feature was inspired by Rust which does precisely this:
+
+- Example rust error message:
+
+   ```
+   error[E0268]: `break` outside of a loop
+    --> src/lib.rs:2:5
+     |
+   2 |     break rust;
+     |     ^^^^^^^^^^ cannot `break` outside of a loop
+   ```
+
+  The error identifier is `E0268` here.
+
+- Index of all errors:
+  https://doc.rust-lang.org/error-index.html
+
+
+## Imaginary example
+
+Current behavior:
+
+```
+# let x = 2 + "foo";;
+Error: This expression has type string but an expression was expected of type
+         int
+```
+
+New behavior:
+
+```
+# let x = 2 + "foo";;
+Error[tyco007]: This expression has type string but an expression
+  was expected of type int
+```
+
+In this example the error identifier is `tyco007`.
+
+## Proposed identifier format
+
+(We are of course happy to change this if there is demand.)
+
+We propose messages such as "tyco007" formed of a "group", the prefix
+"tyco", and a number, 7.
+
+We propose to keeep a fixed length for messages, 7 characters,
+following the Rust approach. (A guess about the motivations: it makes
+formatting more consistent and search easier.) Currently the format we
+use has one group per compiler module emitting errors (hre "tyco" is
+short for "typecore"), but one could move to E123456 later in the
+future using a single group "E".
+
+```ocaml
+type error_identifier = {
+  group: string; (* 4 characters *)
+  num: int; (* 0 <= num < 1000 *)
+}
+```
+
+Note that the format proposed here is very different from the format
+of warning identifiers (`missing-record-field-pattern`). Users want to
+refer to specific warnings in their build system and code to
+enable/disable them, so we want descriptive, human-friendly
+identifiers. In constrat we don't know of a reason to refer to
+a specific error in our code, so error identifiers need not be
+readable and informative; we want them short and to the point.
+
+(Why the group prefix then? Actually it simplifies the implementation
+to give per-module identifiers instead of trying to pick global
+numbers and avoiding conflicts between two errors trying to use the
+same number. It might also help users identify classes of errors.)
```

Comment:
I'm not really convinced by the format nor the rationale. 

I think the fact that it's an abbreviation that maps to the module is likely to be interesting only to compiler devs. And the day your code moves to another module what do you do with the uid ? Is it that complicated to have a single module `Error_id` where you define errors and which you draw from ? 

Why not keep it to EXXX/WXXX, short and actually tells the user whether that's an error or a warning.

Reply:
The module name in the example was mostly a proxy for the class of errors. For instance, extending the examples
- `synt` : syntax errors
- `tyco` : core language type errors
- `tymo `: module languages
- `tymi` module inclusion errors
- `tyex` : type expression errors
- `link`: linking error
- `cpmo` : module compilation error
If we ever change the location/classification of an error, it seems fine to change the identifiers of the errors.

One advantage that I see with those errors is that thematically adjacent errors will share similar identifiers even after many renaming and moves (which doesn't happen that often). Contrarily, with dense set of numerical identifiers, a single splits of errors might move two sibling errors far away from each other.
However, I kind of agree that the "themes" themselves might be only partially understandable at first glance for non-compiler developers, but I think that the similarity of the errors will make sense for users. 

Defining all error datatypes in a single module does not really work that well with the current architecture: error constructors often carry complex payload that are strongly tied to the concepts of the modules that raise them.
We could define the identifier part centrally, and dispatch the error constructors themselves to those identifiers, but that creates two distinct source of truths and probably make a coherency tool a little more mandatory.

Overall, we could definitively have a single numerical identifiers, but the idea of having group of errors seemed slightly attractive both in term of user experience and compiler development. But it could be totally be than my vision of user experience is far too biased in that specific instances.

Reply:
My 2 cents

```
synt : syntax errors
tyco : core language type errors
tymo : module languages
tymi : module inclusion errors
tyex : type expression errors
link : linking error
cpmo : module compilation error
```

This map of knowledge only makes sense when you look at it entirely. When those abbreviations might appear on your console, out of context it wouldn't give any valuable information.

For example when "tyex" appears on your errors for the first time until the N time and realise it has some meaning (and there are other codes that mean other errors) that's already late since you would understand the map as a hole and context isn't needed anymore.

I like the proposal, it adds searchability to errors and given the extended places to ask/receive help online it's huge. I would either make it indexable by a number or give the entire context in the error. Instead of `synt` -> `SYNTAX_ERROR_X` 

PS: Sorry If I jumped into an issue where no one called me out